### PR TITLE
Modify CBMC proofs to make assumptions about malloc explicit

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Issue #, if available:*

N/A.

*Description of changes:*

- Modify CBMC proofs to make assumptions about malloc explicit;
- Upgrade the template version used in CBMC proofs (e.g., new version includes "malloc may fail" flags);
- Revert changes from PR https://github.com/awslabs/aws-c-common/pull/600;
- Adds stub for `s_remove_entry`in `hash_table` proofs;

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by SaswatPadhi — APPROVED** [review]
LGTM! Thanks @feliperodri

File: verification/cbmc/stubs/s_remove_entry_override.c
Comment (feliperodri) [inline verification/cbmc/stubs/s_remove_entry_override.c:16]: I'm tracking this stub with issue https://github.com/awslabs/aws-c-common/issues/709, so we can remove it after improving CBMC performance.

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (danielsn) [inline verification/cbmc/include/proof_helpers/make_common_data_structures.h:24]: I'm not sure we need bounded malloc at all here. CBMC should already handle returning `null` if you ask for too much memory

File: verification/cbmc/proofs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c:16]: Is there a reason you conjoin these rather than having two assumes?

File: verification/cbmc/proofs/aws_hash_iter_delete/Makefile
Comment (danielsn) [inline verification/cbmc/proofs/aws_hash_iter_delete/Makefile:17]: This should be `ABSTRACTIONS` not proof sources.

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (danielsn) [inline verification/cbmc/include/proof_helpers/make_common_data_structures.h:25]: What is this doing here? Shouldn't we do an explicit notnull check?

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (danielsn) [inline verification/cbmc/include/proof_helpers/make_common_data_structures.h:22]: This has specific meaning for `count == 0` and `count > MAX_MALLOC_SIZE`. Document that.

File: verification/cbmc/proofs/aws_array_list_comparator_string/aws_array_list_comparator_string_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_array_list_comparator_string/aws_array_list_comparator_string_harness.c:16]: I assume you've check that these are all preconditions on the actual code.

File: verification/cbmc/proofs/aws_array_list_init_static/aws_array_list_init_static_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_array_list_init_static/aws_array_list_init_static_harness.c:28]: What about the `len == 0, ptr == NULL` case

File: verification/cbmc/stubs/s_remove_entry_override.c
Comment (danielsn) [reply verification/cbmc/stubs/s_remove_entry_override.c:16]: Do we have a proof for this function that has the same pre/post conditions?

File: verification/cbmc/sources/proof_allocators.c
Comment (danielsn) [inline verification/cbmc/sources/proof_allocators.c:58]: The actual allocator goes farther: it has a fatal assert that size is not 0 
https://github.com/awslabs/aws-c-common/blob/master/source/allocator.c#L69

File: verification/cbmc/sources/make_common_data_structures.c
Comment (danielsn) [inline verification/cbmc/sources/make_common_data_structures.c:193]: use `{}` here

File: verification/cbmc/proofs/aws_string_eq_ignore_case/aws_string_eq_ignore_case_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_string_eq_ignore_case/aws_string_eq_ignore_case_harness.c:12]: `ensure` is not the right name anymore, given that it can return `null`

File: verification/cbmc/proofs/aws_hash_table_remove/Makefile
Comment (danielsn) [inline verification/cbmc/proofs/aws_hash_table_remove/Makefile:26]: `ABSTRACTIONS`

File: verification/cbmc/proofs/aws_string_compare/aws_string_compare_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_string_compare/aws_string_compare_harness.c:15]: This `IMPLIES(x != NULL, foo_is_valid(x))` pattern occurs often enough to be worth its own macro

File: verification/cbmc/proofs/aws_string_destroy_secure/aws_string_destroy_secure_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_string_destroy_secure/aws_string_destroy_secure_harness.c:19]: it would be cleaner to have a `aws_string old_string = *str;` and use that.

File: verification/cbmc/proofs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c
Comment (danielsn) [inline verification/cbmc/proofs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c:16]: why the conjunction here?

**Review by danielsn — CHANGES_REQUESTED** [review]

File: verification/cbmc/stubs/s_remove_entry_override.c
Comment (feliperodri) [reply verification/cbmc/stubs/s_remove_entry_override.c:16]: We don't have a proof for this function. I created issue https://github.com/awslabs/aws-c-common/issues/711 to add this on another PR.

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (feliperodri) [reply verification/cbmc/include/proof_helpers/make_common_data_structures.h:24]: We still need bounded_malloc not only to bound the input but also to cover the case when the input is 0 and malloc return a invalid input.

File: verification/cbmc/proofs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c:16]: No reason, I split into two assumes.

File: verification/cbmc/proofs/aws_hash_iter_delete/Makefile
Comment (feliperodri) [reply verification/cbmc/proofs/aws_hash_iter_delete/Makefile:17]: There are many files that use `PROOF_SOURCES` instead of `ABSTRACTIONS`.

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (feliperodri) [reply verification/cbmc/include/proof_helpers/make_common_data_structures.h:25]: I replace it with a nullness check.

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (feliperodri) [reply verification/cbmc/include/proof_helpers/make_common_data_structures.h:22]: I added a comment for it.

File: verification/cbmc/proofs/aws_array_list_comparator_string/aws_array_list_comparator_string_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_array_list_comparator_string/aws_array_list_comparator_string_harness.c:16]: Yes.

File: verification/cbmc/proofs/aws_array_list_init_static/aws_array_list_init_static_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_array_list_init_static/aws_array_list_init_static_harness.c:28]: This maps to an assumption in the function under verification and `len` is not a input in this function.
```c
AWS_STATIC_IMPL
void aws_array_list_init_static(
    struct aws_array_list *AWS_RESTRICT list,
    void *raw_array,
    size_t item_count,
    size_t item_size) {

    AWS_FATAL_PRECONDITION(list != NULL);
    AWS_FATAL_PRECONDITION(raw_array != NULL);
    AWS_FATAL_PRECONDITION(item_count > 0);
    AWS_FATAL_PRECONDITION(item_size > 0);
```

File: verification/cbmc/sources/proof_allocators.c
Comment (feliperodri) [reply verification/cbmc/sources/proof_allocators.c:58]: So I could add an assumption in all allocators to consider `size != 0`.

File: verification/cbmc/sources/make_common_data_structures.c
Comment (feliperodri) [reply verification/cbmc/sources/make_common_data_structures.c:193]: Done.

File: verification/cbmc/proofs/aws_hash_table_remove/Makefile
Comment (feliperodri) [reply verification/cbmc/proofs/aws_hash_table_remove/Makefile:26]: I have an issue to correct this (https://github.com/awslabs/aws-c-common/issues/712);

File: verification/cbmc/proofs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c:16]: Done.

File: verification/cbmc/proofs/aws_string_destroy_secure/aws_string_destroy_secure_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_string_destroy_secure/aws_string_destroy_secure_harness.c:19]: Done.

File: verification/cbmc/proofs/aws_string_eq_ignore_case/aws_string_eq_ignore_case_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_string_eq_ignore_case/aws_string_eq_ignore_case_harness.c:12]: I replaced all instances with `nondet_allocate_string_bounded_length`.

File: verification/cbmc/proofs/aws_string_compare/aws_string_compare_harness.c
Comment (feliperodri) [reply verification/cbmc/proofs/aws_string_compare/aws_string_compare_harness.c:15]: I agree, so I created this issue https://github.com/awslabs/aws-c-common/issues/713 to add this on a separate PR.

File: verification/cbmc/stubs/s_remove_entry_override.c
Comment (danielsn) [reply verification/cbmc/stubs/s_remove_entry_override.c:16]: OK

File: verification/cbmc/include/proof_helpers/make_common_data_structures.h
Comment (danielsn) [reply verification/cbmc/include/proof_helpers/make_common_data_structures.h:25]: We should decide what to do in this case, but I won't block the PR on it.

File: verification/cbmc/sources/proof_allocators.c
Comment (danielsn) [reply verification/cbmc/sources/proof_allocators.c]: Yeah. Won't block this PR on it, but probably a good idea to assert that as an added check

**Review by danielsn — APPROVED** [review]

**Pull request merged**
